kernel/src/log_segment.rs

```diff
@@ -422,27 +424,84 @@ impl LogSegment {
         )?))
     }
 
-    // Do a lightweight protocol+metadata log replay to find the latest Protocol and Metadata in
-    // the LogSegment
+    // Find the latest Protocol and Metadata in the LogSegment. This is done by replaying the log
+    // until the latest_crc_file/latest checkpoint is reached, at which point we can just return the
+    // P+M from the CRC directly or read from checkpoint.
+    //
+    // Note that according to the spec CRC files must contain the latest Protocol and Metadata. If
+    // we encounter a CRC file without P+M, this will return an error (and user should likely just
+    // remove such a file).
+    //
+    // Four steps:
+    // 1. If we have CRC at this version, we can just read the Protocol and Metadata from it.
+    // 2. Else, create a 'pruned' (CRC-rooted) log segment that is only the commits after the
+    //    latest CRC (if there is no CRC, we just use the whole log segment)
+    // 3. Do PM replay on the pruned log segment to find the latest Protocol/Metadata actions.
+    // 4. If no Protocol/Metadata was found, return Protocol/Metadata from CRC (if CRC doesn't
+    //    exist, we are guaranteed to get a Protocol/Metadata from checkpoint in (3)).
+    //
+    // Note that there is an optimization here: we _could_ have different types of log segments.
+    // Instead of a log segment always containing all metadata for the table, we could have a
+    // "PM-LogSegment" that is only guaranteed to contain Protocol and Metadata actions. Basically,
+    // a "PM-LogSegment" would be a log segment that is checkpoint OR CRC-rooted, whereas a full
+    // LogSegment must be checkpoint-rooted (or start from 0).
     pub(crate) fn protocol_and_metadata(
         &self,
         engine: &dyn Engine,
     ) -> DeltaResult<(Option<Metadata>, Option<Protocol>)> {
-        let actions_batches = self.replay_for_metadata(engine)?;
+        // 1. simple case if we have CRC at this version
+        if let Some(latest_crc_file) = &self.latest_crc_file {
+            if self.end_version == latest_crc_file.version {
+                let path = &latest_crc_file.location;
+                let pm = ProtocolMetadata::try_from_crc(path.clone(), engine)?;
+                return Ok((Some(pm.metadata), Some(pm.protocol)));
+            }
+        }
+
+        // 2. construct a "protocol-metadata log segment"
+        //    a) if we have a CRC file, we only need commits (CRC_version, target_version]
+        //    b) if we don't have a CRC, we proceed as normal
+        let protocol_metadata_log_segment: Cow<'_, Self> =
+            if let Some(latest_crc_file) = &self.latest_crc_file {
+                let mut log_segment = self.clone();
+                log_segment.checkpoint_version = None;
```

Comment:
Have we pruned here so we know that any crc has to be _after_ the checkpoint?

Reply:
ah this is a good find - I don't think we have. so we _could_ have a CRC which is before the checkpoint. I'll fix in the short-term by just checking here that CRC version > checkpoint version. (or >=? should we prefer to use CRC if both are present? - I can do some looking around).

Though I wonder if we should fix this here or modify our `LogSegment` semantics to only track the latest CRC if it's after the most recent checkpoint? Would there be a case where you'd want to use a CRC file if it's before a checkpoint? I suppose CRC would be much smaller than checkpoint in the common case so _could_ speed up the raw file fetch if you're just doing PM query? Though we wouldn't expect much actual 'read' speedup since parquet can column prune, right?